Sources/Profile-Modify.php

```diff
@@ -2495,10 +2495,52 @@ function alert_count($memID, $unread = false)
 	}
 
 	// Now check alerts again and remove any they can't see.
+	$deletes = array();
+	$num_unread_deletes = 0;
 	foreach ($alerts as $id_alert => $alert)
 	{
 		if (!$alert['visible'])
+		{
+			if (empty($alert['is_read']))
+				$num_unread_deletes++;
+
 			unset($alerts[$id_alert]);
+			$deletes[] = $id_alert;
+		}
+	}
+
+	// Penultimate task - delete these orphaned, invisible alerts, otherwise they might hang around forever.
+	// This can happen if they are deleted or moved to a board this user cannot access.
+	// Note that unread alerts are never purged.
+	if (!empty($deletes))
+	{
+		$smcFunc['db_query']('', '
+			DELETE FROM {db_prefix}user_alerts
+			WHERE id_alert IN ({array_int:alerts})',
+			array(
+				'alerts' => $deletes,
+			)
+		);
+	}
+
+	// One last thing - tweak counter on member record.
+	// Do it directly, as updateMemberData() calls this function, and may create a loop.
+	// Note that $user_info is not populated when this is invoked via cron, hence the CASE statement.
+	if ($num_unread_deletes > 0)
+	{
+		$smcFunc['db_query']('', '
+			UPDATE {db_prefix}members
+			SET alerts =
+				CASE
+					WHEN alerts < {int:unread_deletes} THEN 0
+					ELSE alerts - {int:unread_deletes}
+				END
+			WHERE id_member = {int:member}',
+			array(
+				'unread_deletes' => $num_unread_deletes,
```

Comment:
`$num_unread_deletes` is never incremented above, so it will always be 0. I think you can simply get rid of this variable and instead use `count($deletes)` on this line.

Reply:
It's incremented on L2505 above.

And it tests fine - this fix has been on my production forum for quite some time.

Note I have a count validation script that runs occasionally to ensure the counts don't fall out of sync again - they aren't.

Reply:
One other note: unread deletes <> deletes, so the count won't work.

Reply:
Derp. I missed that somehow.